### PR TITLE
Feature: ioc

### PR DESCRIPTION
**Pull request by eapg**
* Adding Ioc 
* Adding Ioc unit test
* Singleton decorator added, but will not be used for the moment.

### CODE REVIEW

File: src/core/ioc.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+from src.core.singleton_decorator import singleton
+from src.lib.repositories.impl import *
+from src.api.controllers import *
+
+
+@singleton
+class Ioc:
+
+    _instance_ioc = {}
```
Comment (reynoldmorel) [inline src/core/ioc.py:11]: Could initialize this within `__init__`?

File: src/core/ioc.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    def __init__(self):
+
+        self._init_repositories()
+        self._init_controllers()
+
+    def _init_repositories(self):
+        self._instance_ioc[
+            "product_repository_impl"
```
Comment (reynoldmorel) [inline src/core/ioc.py:20]: It is not a good idea to set `impl` at the end. You should only have 1 source of true for all the registered instances in the pool.

File: src/core/ioc.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# structure for the ioc to have just one instance
+
+from src.core.singleton_decorator import singleton
+from src.lib.repositories.impl import *
+from src.api.controllers import *
+
+
+@singleton
+class Ioc:
```
Comment (reynoldmorel) [inline src/core/ioc.py:9]: Also, could export a function that will build single instance for you?

File: src/core/ioc.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        self._instance_ioc[
+            "order_detail_controller"
+        ] = order_detail_controller.OrderDetailController(
+            self._instance_ioc["order_detail_repository_impl"]
+        )
+        self._instance_ioc["chef_controller"] = chef_controller.ChefController(
+            self._instance_ioc["chef_repository_impl"]
+        )
+
+    def put_instance(self, instance, instance_id):
```
Comment (reynoldmorel) [inline src/core/ioc.py:72]: You should protect developers from override of your registered instance, either verify an existing instance can not be override it from the outside or hide `put_instance`

File: src/core/singleton_decorator.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# structure for the singleton decorator
+
+
+def singleton(cls):
+
+    instance = [None]
```
Comment (reynoldmorel) [inline src/core/singleton_decorator.py:6]: Why are you using `None` for this array?

File: src/core/singleton_decorator.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+    def wrapper(*args, **kwargs):
```
Comment (reynoldmorel) [inline src/core/singleton_decorator.py:8]: Nice code. Could you explain what is happening here?

File: src/tests/core/ioc_test.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+class IocTestCase(unittest.TestCase):
+    def test_instance_not_none(self):
+
+        ioc_instance = Ioc()
+
+        product_controller_instance = ioc_instance.get_instance("product_controller")
+        self.assertIsNotNone(product_controller_instance)
+
+    def test_same_instance(self):
```
Comment (reynoldmorel) [inline src/tests/core/ioc_test.py:14]: Could you check instance 1 and 2 are also de same?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/core/ioc.py (hunk 1 above)
Comment (eapg) [reply src/core/ioc.py:11]: No, if i do that the variable will belong to the object instance not to the class and to be singleton i want the varible to belong the class thus the variable will be share for all instances.

File: src/core/ioc.py (hunk 2 above)
Comment (eapg) [reply src/core/ioc.py:20]: copied!

File: src/core/singleton_decorator.py (hunk 5 above)
Comment (eapg) [reply src/core/singleton_decorator.py:6]: It is to check in the code below if the instance[0] is None, thus i know that i'm creating the instance for the firts time.
it is the same to put instance[0] = None

File: src/core/ioc.py (hunk 4 above)
Comment (eapg) [reply src/core/ioc.py:72]: done, i removed the put_instance method

File: src/tests/core/ioc_test.py (hunk 7 above)
Comment (eapg) [reply src/tests/core/ioc_test.py:14]: done!

File: src/core/ioc.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        )
+
+    def get_instance(self, instance_id):
+        return self._instance_ioc[instance_id]
+
+
+instance_ioc = Ioc()
+
+
+def get_instance_ioc(instance_id):
```
Comment (reynoldmorel) [inline src/core/ioc.py:75]: Could make this return only `ioc_instance`?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/core/decorators/singleton_decorator.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# structure for the singleton decorator
+
+
+def singleton(cls):
+
+    instance = {}
+
+    def wrapper(*args, **kwargs):
```
Comment (reynoldmorel) [inline src/core/decorators/singleton_decorator.py:8]: Can we use  class name instead as the key? There is no need yo use the whole class as the key of the dict

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/core/ioc.py (hunk 8 above)
Comment (eapg) [reply src/core/ioc.py]: Done

**Pull request merged**